### PR TITLE
Added support for Common Function Attributes

### PR DESCRIPTION
**Pull request by shubhamnarlawar77**
The pull request(done under guidance of Martin Liska @marxin) generates common function attributes during generation phase of functions and outputs during output phase.  

Mentioning @jxyang and @andikleen also in this pull request.

Please review and suggest changes.

### CODE REVIEW

File: src/Attribute.h [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	BooleanAttribute(string, int);
+	string OutputAttributes();
+};
+
+//Generates Multi Valued attributes such as visibility("option"), no_sanitize("option"), etc
+class MultiValuedAttribute : public Attribute
+{
+public:
+	//stores various options of attributes e.g. visibility options - default, hidden, internal and protected
+	vector<string> attribute_values;
```
Comment (jxyang) [inline src/Attribute.h]: Rename to "choices"?

File: src/Attribute.cpp [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+#ifdef WIN32
+#pragma warning(disable : 4786)   /* Disable annoying warning messages */
+#endif
+
+#include "random.h"
+#include "OutputMgr.h"
+#include "Attribute.h"
+
+void
+AttributeGenerator::GenerateAllAttributes(std::ostream &out)
```
Comment (jxyang) [inline src/Attribute.cpp]: The attributes are for functions only as of now? Maybe rename this to "FunctionAttribute::Output"

File: src/Attribute.cpp [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+{
+}
+
+BooleanAttribute::BooleanAttribute(string name, int prob)
+	: Attribute(name, prob)
+{
+}
+
+string
+BooleanAttribute::OutputAttributes()
```
Comment (jxyang) [inline src/Attribute.cpp]: Rename to "make_random" to be consistent with other language constructs.

File: src/Attribute.cpp [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+		return "";
+}
+
+MultiValuedAttribute::MultiValuedAttribute(string name, int prob, vector<string> arguments)
+	: Attribute(name, prob), attribute_values(arguments)
+{
+}
+
+string
+MultiValuedAttribute::OutputAttributes()
```
Comment (jxyang) [inline src/Attribute.cpp]: -> "make_random"

File: src/Attribute.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+class Attribute
+{
+public:
+	//Name of attribute
+	string attribute;
+	//Attribute generation probability
+	int attribute_probability;
+	Attribute(string, int);
+	//Checks attribute probability and generate is accordingly
+	virtual string OutputAttributes() = 0;
```
Comment (jxyang) [inline src/Attribute.h]: -> "make_random"

File: src/Attribute.h [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+{
+public:
+	//stores various options of attributes e.g. visibility options - default, hidden, internal and protected
+	vector<string> attribute_values;
+	MultiValuedAttribute(string, int, vector<string>);
+	string OutputAttributes();
+};
+
+//Generates function and types attributes
+class AttributeGenerator
```
Comment (jxyang) [inline src/Attribute.h]: Rename to "FunctionAttribute"

File: src/Attribute.h [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+{
+public:
+	//stores instances of Attribute
+	vector<Attribute*> attributes;
+	void GenerateAllAttributes(std::ostream &);
```
Comment (jxyang) [inline src/Attribute.h]: Rename to "Output“

File: src/Function.cpp [hunk 8]
```diff
@@ -66,17 +66,36 @@
...
 
+AttributeGenerator func_attr_generator;
 static vector<Function*> FuncList;		// List of all functions in the program
 static vector<FactMgr*>  FMList;        // list of fact managers for each function
 static long cur_func_idx;				// Index into FuncList that we are currently working on
 static bool param_first=true;			// Flag to track output of commas
 static int builtin_functions_cnt;
 
+void
+Function::GenerateAttributes()
```
Comment (jxyang) [inline src/Function.cpp]: Rename to "InitializeAttributes"

File: src/Attribute.h (hunk 6 above)
Comment (shubhamnarlawar77) [reply src/Attribute.h]: The class will be used to generate function, type, variable and label attributes later. Is it fine if I keep class name as `AttributeGenerator` ?

File: src/Attribute.h [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+using namespace std;
+
+class AttributeGenerator;
+
+//Base class with pure virtual methods which will be derived as per different types of attribute generation
+class Attribute
+{
+public:
+	//Name of attribute
+	string attribute;
```
[the remark this answers is not in the document]
Comment (shubhamnarlawar77) [reply src/Attribute.h]: Done

File: src/Attribute.h [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+	//Attribute generation probability
+	int attribute_probability;
```
[the remark this answers is not in the document]
Comment (shubhamnarlawar77) [reply src/Attribute.h]: Done

File: src/Attribute.h [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+//Generates Boolean attributes such as hot, cold, used, unused, deprecated, etc
+class BooleanAttribute : public Attribute
+{
+public:
+	BooleanAttribute(string, int);
+	string OutputAttributes();
+};
+
+//Generates Multi Valued attributes such as visibility("option"), no_sanitize("option"), etc
+class MultiValuedAttribute : public Attribute
```
[the remark this answers is not in the document]
Comment (shubhamnarlawar77) [reply src/Attribute.h]: Done.

File: src/Attribute.h (hunk 1 above)
Comment (shubhamnarlawar77) [reply src/Attribute.h]: Done.

File: src/Attribute.cpp (hunk 2 above)
Comment (shubhamnarlawar77) [reply src/Attribute.cpp]: Yes. But keeping it as `AttributeGenerator` seems appropriate as I want to send type, variable and label attributes also.

Changed `GenerateAllAttributes` to `Output`.

File: src/Attribute.cpp (hunk 3 above)
Comment (shubhamnarlawar77) [reply src/Attribute.cpp]: Sure. Done

File: src/Attribute.cpp (hunk 4 above)
Comment (shubhamnarlawar77) [reply src/Attribute.cpp]: Done

File: src/Attribute.h (hunk 5 above)
Comment (shubhamnarlawar77) [reply src/Attribute.h]: Done

File: src/Attribute.h (hunk 7 above)
Comment (shubhamnarlawar77) [reply src/Attribute.h]: Done.

File: src/Function.cpp (hunk 8 above)
Comment (shubhamnarlawar77) [reply src/Function.cpp]: Done. Please check it!!

**Comment by jxyang** [comment]
Looks good. Please fix the build error. I'd write the initializer as:
```
vector<string> common_func_attributes = vector<string>{"artificial", "flatten", ...};
```

**Comment by shubhamnarlawar77** [comment]
> Looks good. Please fix the build error. I'd write the initializer as:
> 
> ```
> vector<string> common_func_attributes = vector<string>{"artificial", "flatten", ...};
> ```
Doing it still do not solve build error. Can Travis-ci version be updated? It seems to use version < c++11 now.

One of way to fix build error is to push one string at a time in vector `common_func_attributes` but it seems inappropriate ?

**Comment by jxyang** [comment]
We should find the right syntax rather than changing the CI script.

Try `vector<string> common_func_attributes{ "artificial", "flatten"};`.

**Comment by shubhamnarlawar77** [comment]
Sure. I will try other approaches. The  build failed for above one also.

Could you name some more known approaches if possible?

**Comment by shubhamnarlawar77** [comment]
> We should find the right syntax rather than changing the CI script.
> 
> Try `vector<string> common_func_attributes{ "artificial", "flatten"};`.

Now, I have created a function `InitializeAttributeChoices()`. I guess build is done with option -std=c++03 and there are no string initializer in c++03, hence I have followed this approach. Is this fine?

**Comment by jxyang** [comment]
@eeide Can you help updating the compiler version?

@shubhamnarlawar77 Thanks for the contribution.

**Comment by eeide** [comment]
Since this issue is merged, I'm going to assume that "updating the compiler version" has been done.

**Comment by jxyang** [comment]
No it hasn't. We worked around the problem with a function to initialize vector of strings.

**Comment by eeide** [comment]
@jxyang, can you make a new issue that describes what you think should be done about "updating the compiler version"? Thanks!

**Pull request merged**